### PR TITLE
#84 and #110

### PR DESCRIPTION
**Pull request by yasserf**

### CODE REVIEW

File: src/record/record-handler.js [hunk 1]
```diff
@@ -106,6 +120,47 @@ RecordHandler.prototype.handle = function( socketWrapper, message ) {
...
+ * @param   {SocketWrapper} socketWrapper the socket that send the request
+ * @param   {Object} message parsed and validated message
+ *
+ * @private
+ * @returns {void}
+ */
+RecordHandler.prototype._hasRecord = function( socketWrapper, message ) {
+	var recordName = message.data[ 0 ],
+		onComplete = function( record ) {
+			var hasRecord = record ? 'T' : 'F';
```
Comment (WolframHempel) [inline src/record/record-handler.js]: Could we use `C.TYPES.TRUE` and C.`TYPES.FALSE`

Comment (yasserf) [inline src/record/record-handler.js]: Yup!

File: src/record/record-handler.js [hunk 2]
```diff
@@ -106,6 +120,47 @@ RecordHandler.prototype.handle = function( socketWrapper, message ) {
...
+			socketWrapper.sendMessage( C.TOPIC.RECORD, C.ACTIONS.HAS, [ recordName, hasRecord ] );
+		};
+
+	new RecordRequest( recordName, this._options, socketWrapper, onComplete.bind( this ) );
```
Comment (WolframHempel) [inline src/record/record-handler.js]: Do we rely on the request timing out if an error occurred? Maybe rather pass an error callback.

File: src/record/record-handler.js [hunk 3]
```diff
@@ -106,6 +120,47 @@ RecordHandler.prototype.handle = function( socketWrapper, message ) {
...
+ * @private
+ * @returns {void}
+ */
+RecordHandler.prototype._snapshot = function( socketWrapper, message ) {
+	var recordName = message.data[ 0 ],
+		onComplete = function( record ) {
+			if( record ) {
+				this._sendRecord( recordName, record, socketWrapper );
+			} else {
+				//TODO: This needs to be improved
```
Comment (WolframHempel) [inline src/record/record-handler.js]: Agreed. Maybe we should rather send an error than introducing a new concept of null for missing

File: src/record/record-handler.js [hunk 4]
```diff
@@ -176,6 +231,20 @@ RecordHandler.prototype._create = function( recordName, socketWrapper ) {
  */
 RecordHandler.prototype._read = function( recordName, record, socketWrapper ) {
 	this._subscriptionRegistry.subscribe( recordName, socketWrapper );
+	this._sendRecord( recordName, record, socketWrapper );
+};
+
+/**
+ * Sends the records data current data once done
```
Comment (WolframHempel) [inline src/record/record-handler.js]: Hm - what changed here?

File: src/record/record-handler.js (hunk 3 above)
Comment (yasserf) [inline src/record/record-handler.js]: Is it an error though? Its correct to request something that doesn't exist

File: test/record/record-handlerSpec.js [hunk 5]
```diff
@@ -62,6 +62,51 @@ describe( 'record handler handles messages', function(){
 		expect( clientA.socket.lastSendMessage ).toBe( msg( 'R|R|existingRecord|3|{"firstname":"Wolfram"}+' ));
 	});
 
+	it( 'returns true for HAS if message exists', function(){
+		recordHandler.handle( clientA, {
+			raw: msg( 'R|H|existingRecord' ),
+			topic: 'RECORD',
```
Comment (WolframHempel) [inline test/record/record-handlerSpec.js]: topic: 'R'

File: src/record/record-handler.js (hunk 4 above)
Comment (yasserf) [inline src/record/record-handler.js]: Read now is a combination of subscribe and snapshot

File: test/record/record-handlerSpec.js [hunk 6]
```diff
@@ -62,6 +62,51 @@ describe( 'record handler handles messages', function(){
...
+		recordHandler.handle( clientA, {
+			raw: msg( 'R|H|nonExistingRecord' ),
+			topic: 'RECORD',
+			action: 'H',
+			data: [ 'nonExistingRecord' ]
+		});
+
+		expect( clientA.socket.lastSendMessage ).toBe( msg( 'R|H|nonExistingRecord|F+' ) );
+	});
+
```
Comment (WolframHempel) [inline test/record/record-handlerSpec.js]: Could we add a test for a failing record request

File: test/record/record-handlerSpec.js (hunk 5 above)
Comment (yasserf) [inline test/record/record-handlerSpec.js]: Good catch, I think I copied that from somewhere so will need to take a look where

File: src/record/record-handler.js (hunk 2 above)
Comment (yasserf) [inline src/record/record-handler.js]: Done

File: test/record/record-handlerSpec.js (hunk 6 above)
Comment (yasserf) [inline test/record/record-handlerSpec.js]: Done

**Pull request merged**